src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -105,6 +106,17 @@ public partial class Ping
                     {
                         fixed (byte* bytesPtr = &receiveBuffer[0])
                         {
+                            if (isIpv4)
+                            {
+                                // Determine actual size of IP header
+                                byte ihl = (byte)(bytesPtr[0] & 0x0f); // Internet Header Length
+                                ipHeaderLength = 4 * ihl;
+                                if (bytesReceived - ipHeaderLength < IcmpHeaderLengthInBytes)
+                                {
+                                    continue; // Not enough bytes to reconstruct actual IP header + ICMP header.
```

Comment:
I also had similar concerns when looking into this code first time, however I'm no expert in network and async I/O, so I just relied on the previous logic.

Thank you for the review and hope someone can improve this later :)